File: benchmarks/cpp/loops/branch_loop.cpp
#include <iostream>
#include <chrono>
using namespace std;
using namespace chrono;

int main() {
    const long long N = 1000000000;
    volatile long long sum = 0;  // prevents over-optimization

    auto start = high_resolution_clock::now();

    for (long long i = 1; i <= N; i++) {
        if (i % 2 == 0)
            sum += 1;
        else
            sum -= 1;
    }

    auto end = high_resolution_clock::now();
    duration<double> elapsed = end - start;

    cout << "C++ Branch Loop Time: " << elapsed.count() << " seconds\n";
    return 0;
}

File: benchmarks/rust/loops/branch_loop.rs
use std::time::Instant;

fn main() {
    let n: i64 = 1_000_000_000;
    let mut _sum: i64 = 0;

    let start = Instant::now();

    for i in 1..=n {
        if i % 2 == 0 {
            _sum += 1;
        } else {
            _sum -= 1;
        }
    }

    let elapsed = start.elapsed();
    println!("Rust Branch Loop Time: {:.6} seconds", elapsed.as_secs_f64());
}
